---
title: "ChIPUtils: Quick start"
author: |
        | Rene Welch and S&#252;nd&#252;z Kele&#351;
        | Department of Statistics, University of Wisconsin-Madison
output: 
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Vignette Title}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval = TRUE,
                      message = FALSE,
                      warning = FALSE)

library(tidyverse)
library(ChIPUtils)

theme_set(theme_bw())

```

# What is ChIPUtils

**ChIPUtils** is a package for very general ChIP-seq analysis. The basic idea behind this package is to gather several of the common procedures we use when starting to analize a new ChIP dataset. To load the package is necessary to use: 

```{r load,include=TRUE}

library(ChIPUtils)

```

For the vignette we use the following samples which are based on ENCODE datasets:

```{r files,echo = FALSE}

bamfiles = list.files(system.file("extdata",package = "ChIPUtils"),
                   full.names = TRUE,recursive = TRUE,
                   pattern = "bam")
bamfiles = bamfiles %>% {.[-grep("bai",.)]} 
bamfiles %>% basename()

```

## **ChIPdata**: ChIPUtils base structure

To use the different methods that the package contains, it is neccesary to create a **ChIPdata** object, for which is necessary to have a file in bam format with aligned reads and to indicate if the reads are single or paired ended:

```{r chipdata}

example = ChIPdata(bamfiles[1],isPE = FALSE)
example %>% slotNames()
 
```

and the number of aligned reads in the experiment can be retrieved by:

```{r nreads}
nreads(example)
```

# Quality Control (QC) metrics

Using **ChIPUtils** we can calculate the PBC and SCC, which are the two most commonly used QC metrics for [ChIP-seq](https://www.encodeproject.org/data-standards/2012-quality-metrics/):

```{r pbc,include=TRUE}
PBC(example)
```

```{r scc , include=TRUE,fig.width=5,fig.height=3}
scc = SCC(example,shift = seq(1,300,by = 10),verbose = FALSE)
fl = scc[which.max(scc$corr),]$shift
rl = reads(example)[1] %>% 
  width()
scc %>% 
  ggplot(aes(shift,corr))+geom_line(colour = "red")+
  geom_vline(xintercept = fl,linetype = 2)
```

Using the SCC curve above, we can calculate the Normalized Strand Cross-Correlation by:

```{r nsc,include=TRUE}
NSC(scc)
```

and the Relative Strand Cross-Correlation by:

```{r rsc,include=TRUE}
RSC(scc,rl)
```
 
# Bin exploration

**ChIPUtils** contains a collection of genome sizes, which can be seen by using:

```{r poss,include=TRUE}
possibleChrSizes()
```

For example after picking the `r "hg19"` genome, we can partition it into fixed length bins by using:

```{r bins1 , include=TRUE}
data("chromSizes",package = "ChIPUtils")
createBins(1e4,chrom = chromSizes[["hg19"]]) 

```

Furthermore, we can use the same command to count the number of extended reads that overlap each bin, by adding a **ChIPdata** object and specifying the unobserved fragment length `r fl`:

```{r bins2,include=TRUE}
## we are using only the first three chromosomes 
bins = createBins(1e3,chipdata = example,chrom = chromSizes[["hg19"]][1:3],fragLen = fl)
bins
```

```{r bins3,include=TRUE,fig.width=5,fig.height=3}
tibble(tagCounts = bins$tagCounts) %>% 
  filter(tagCounts > 0) %>% 
  ggplot(aes(tagCounts))+geom_histogram(fill = "white",colour = "black")
```



